src/ddsrt/src/library/posix/library.c

```diff
@@ -0,0 +1,127 @@
+/*
+ * Copyright(c) 2006 to 2018 ADLINK Technology Limited and others
+ *
+ * This program and the accompanying materials are made available under the
+ * terms of the Eclipse Public License v. 2.0 which is available at
+ * http://www.eclipse.org/legal/epl-2.0, or the Eclipse Distribution License
+ * v. 1.0 which is available at
+ * http://www.eclipse.org/org/documents/edl-v10.php.
+ *
+ * SPDX-License-Identifier: EPL-2.0 OR BSD-3-Clause
+ */
+#include <stdio.h>
+#include <dlfcn.h>
+#include <assert.h>
+#include <string.h>
+#include "dds/ddsrt/heap.h"
+#include "dds/ddsrt/library.h"
+
+
+dds_retcode_t
+ddsrt_dlopen(
+    const char *name,
+    bool translate,
+    ddsrt_lib *handle)
+{
+  dds_retcode_t retcode = DDS_RETCODE_OK;
+
+
+  if ( name == NULL || handle == NULL ){
+      retcode = DDS_RETCODE_BAD_PARAMETER;
+
+  } else {
+      *handle = NULL;
+      dlerror();    /* Clear any existing error */
+      if ( (translate) && (strrchr(name, '/') == NULL)) {
+        /* Add lib and suffix to the name and try to open. */
+#if __APPLE__
+        static const char suffix[] = ".dylib";
+#else
+        static const char suffix[] = ".so";
+#endif
+        char* libName = ddsrt_malloc(3 + strlen(name) + sizeof(suffix));
+        sprintf(libName, "lib%s%s", name, suffix);
+        *handle = dlopen(libName, RTLD_GLOBAL | RTLD_NOW);
+        ddsrt_free(libName);
+
+
+      }
+
+      if (*handle == NULL && retcode == DDS_RETCODE_OK) {
+        /* name contains a path,
+         * (auto)translate is disabled or
+         * dlopen on translated name failed. */
+        *handle = dlopen(name, RTLD_GLOBAL | RTLD_NOW);
+
+      }
+
+      if ( *handle != NULL ){
+          retcode = DDS_RETCODE_OK;
+      } else {
+          retcode = DDS_RETCODE_ERROR;
+      }
+  }
+
+  return retcode;
+}
+
+
+dds_retcode_t
+ddsrt_dlclose(
+    ddsrt_lib handle)
+{
+    dds_retcode_t retcode;
+
+    if( handle == NULL ){
+        retcode = DDS_RETCODE_BAD_PARAMETER;
+    }
+    else{
+        retcode = (dlclose(handle) == 0) ? DDS_RETCODE_OK : DDS_RETCODE_ERROR;
+    }
+
+    return retcode;
+}
+
+dds_retcode_t
+ddsrt_dlsym(
+    ddsrt_lib handle,
+    const char *symbol,
+    void **address)
+{
+    dds_retcode_t retcode = DDS_RETCODE_OK;
+
+    if ( symbol == NULL || handle == NULL || address == NULL){
+        retcode = DDS_RETCODE_BAD_PARAMETER;
+    } else {
+        *address =  dlsym(handle, symbol);
+        if( *address == NULL ){
+            retcode = DDS_RETCODE_ERROR;
+        }
+    }
+
+    return retcode;
+}
+
+dds_retcode_t
+ddsrt_dlerror(
+    char *buf,
+    size_t buflen)
+{
+
+    const char *err;
+    dds_retcode_t retcode = DDS_RETCODE_OK;
+
+    if( buf == NULL ){
```

Comment:
An `assert` would be better in this case.

Reply:
As we provide these functions to other components, we decided to convert asserts to NULL checks.
If you think that it is impossible to get buf == NULL, I can change it.

Reply:
I don't think @k0ekk0ek 's point is that it is impossible to get a null pointer for buf; it is rather that there is a good argument to make consider it illegal to do so, i.e., that a precondition is that buf != NULL. That means an assert, preferably a compiler annotation that it must not be a null pointer. The only callers will be internal and there seems to be no good reason to ever call it with a null pointer for buf. Checking for a null pointer will only hide errors.

Reply:
I've changed all parameter checks to assert. We need an assert policy in the coding guideline. (Of course, we need a guideline first)